clients: Add rate limit handling to catalyst client

victorges opened the pull request:
Simple logic, just waiting 15 seconds between each attempt. Should be fine since the request is also really cheap on catalyst side when it gets rate limited, and there will be at most 5 of these looping per task runner instance (3 instances)

My only concern now is that if we're operating at full capacity, these 5 executions will also block handling of any events from the currently running jobs. So we might stop checking if we're getting catalyst callbacks and handling the result callbacks (they become events as well).

Should consider reusing the "delayed task" abstraction used for checking catalyst callbacks for this as well, as a pre-step on the upload task execution in case it got rate limited.

Edit: Fixed the above and implemented an async back off through the delayed event.

victorges committed:
clients: Add rate limit handling to catalyst client

victorges committed:
clients: Add a limit in number of retries

victorges committed:
task/upload: Make sure to backoff after a few retries

thomshutt commented on clients/catalyst.go:
```diff
@@ -70,15 +77,30 @@ func (c *catalyst) UploadVOD(ctx context.Context, upload UploadVODRequest) error
+			return err
+		}
+
+		if attempt >= maxAttempts {
```
So are we giving up if we get rate limited more than 3 times?

victorges replied:
no! we will retry asynchronously instead through a delayed RabbitMQ event. Check the logic on `upload.go`

victorges committed:
clients: Retry exponentially instead

Total retry time will be shorter, 14s of cumulative sleep,
which seems better overall: try 4 times in 15 seconds, then
backoff asynchronously for a minute.

victorges committed:
clients: Make sure we close the connection after each attempt
